replica/tablets.cc

```diff
@@ -308,4 +310,281 @@ future<std::vector<canonical_mutation>> read_tablet_mutations(seastar::sharded<r
+    auto ret = make_lw_shared<sstable_list>();
+    for_each_sstable_set_until(query::full_partition_range, [&] (lw_shared_ptr<sstables::sstable_set> set) {
+        set->for_each_sstable([&] (const sstables::shared_sstable& sst) {
+            ret->insert(sst);
+        });
+        return stop_iteration::no;
+    });
+    return ret;
+}
+
+stop_iteration tablet_sstable_set::for_each_sstable_until(std::function<stop_iteration(const sstables::shared_sstable&)> func) const {
+    return for_each_sstable_set_until(query::full_partition_range, [func = std::move(func)] (lw_shared_ptr<sstables::sstable_set> set) {
+        return set->for_each_sstable_until(func);
+    });
+}
+
+future<stop_iteration> tablet_sstable_set::for_each_sstable_gently_until(std::function<future<stop_iteration>(const sstables::shared_sstable&)> func) const {
+    return for_each_sstable_set_gently_until(query::full_partition_range, [func = std::move(func)] (lw_shared_ptr<sstables::sstable_set> set) {
+        return set->for_each_sstable_gently_until(func);
+    });
+}
+
+bool tablet_sstable_set::insert(sstables::shared_sstable sst) {
+    throw_with_backtrace<std::bad_function_call>();
+}
+bool tablet_sstable_set::erase(sstables::shared_sstable sst) {
+    throw_with_backtrace<std::bad_function_call>();
+}
+
+class tablet_incremental_selector : public sstables::incremental_selector_impl {
+    const tablet_sstable_set& _tset;
+
+    // _cur_set and _cur_selector contain a snapshot
+    // for the currently selected compaction_group.
+    lw_shared_ptr<sstables::sstable_set> _cur_set;
+    std::optional<sstables::sstable_set::incremental_selector> _cur_selector;
+    dht::token _lowest_next_token = dht::maximum_token();
+
+public:
+    tablet_incremental_selector(const tablet_sstable_set& tset)
+            : _tset(tset)
+    {}
+
+    virtual std::tuple<dht::partition_range, std::vector<sstables::shared_sstable>, dht::ring_position_ext> select(const dht::ring_position_view& pos) override {
+        // Always return minimum singular range, such that incremental_selector::select() will always call this function,
+        // which in turn will find the next sstable set to select sstables from.
+        const dht::partition_range current_range = dht::partition_range::make_singular(dht::ring_position::min());
+
+        // pos must be monotonically increasing in the weak sense
+        // but caller can skip to a position outside the current set
+        auto token = pos.token();
+        if (!_cur_set || pos.token() >= _lowest_next_token) {
+            auto idx = _tset.group_of(token);
+            if (!token.is_maximum()) {
+                _cur_set = _tset._sstable_sets[idx];
+            }
+            // Set the next token to point to the possible storage group.
+            // It will be considered later on when the _cur_set is exhausted
+            _lowest_next_token = find_lowest_next_token(_tset.first_token_of(idx+1));
+        }
```

Comment:
what if we're on the last tablet? can it happen idx+1 results in out of bound access? maybe find_lowest_next_token() should receive idx, rather than token. and if we're on last tablet (_tmap.last_tablet() == idx), then max token is returned.

Reply:
It's prepared for that:
```c++
        if (idx >= _sstable_sets.size()) {
            return dht::maximum_token();
        }
```

Reply:
where is that? I cannot find it.

Reply:
maybe it's uncommitted?

Reply:
I used to have logic for that but when I change the selector to call the tablet_map `first_token_of` this protection is now gone.
I'll fix in the next version.